src/hooks/useRequest.ts

```diff
@@ -0,0 +1,90 @@
+import { useCallback, useEffect, useMemo, useRef, useState } from "react";
+
+// Possible utility types for re-use
+type PromiseFn = (...args: any[]) => Promise<any>;
+type Awaited<T> = T extends PromiseLike<infer U> ? Awaited<U> : T;
+type AwaitedReturn<T extends PromiseFn> = Awaited<ReturnType<T>>;
+
+export interface UseRequestOptions<T> {
+  initialValue?: T;
+  onError?: (err: Error) => void;
+  onSuccess?: (res: T) => void;
+}
+
+export interface UseRequestResponse<T> {
+  loading: boolean;
+  error: Error | undefined;
+  data: T | undefined;
+}
+
+export type UseRequestReturn<T extends (...args: any[]) => Promise<T>> = [
+  (...args: Parameters<T>) => Promise<void>,
+  UseRequestResponse<AwaitedReturn<T>>
+];
+
+/**
+ * Provides an API to statefully interact with promises. This hook
+ * prevents state updates if promise has not resolved before component unmounts.
+ * It will return a tuple with a request() function, and an object representing the state
+ */
+export function useRequest<T extends PromiseFn>(
```

Comment:
B/c usage examples will be implemented in the follow-up to this work, I'll provide a quick example below:

```tsx
function UserInfo({ name }: { name: string }) {
  const [requestUserInfo, { data, loading, error }] = useRequest(fetchUserByName);

  useEffect(() => {
    void requestUserInfo(name);
  }, [name]);

  if (loading) return <Loader />;

  if (error) return <Error error={error} />;

  return <UserDetails user={data} />;
}
```